org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/handlers/IndexUtils.java

```diff
@@ -0,0 +1,144 @@
+import org.eclipse.jdt.internal.core.JavaProject;
+import org.eclipse.jdt.internal.core.index.FileIndexLocation;
+import org.eclipse.jdt.internal.core.index.Index;
+import org.eclipse.jdt.internal.core.index.IndexLocation;
+import org.eclipse.jdt.internal.core.search.indexing.IndexManager;
+import org.eclipse.jdt.internal.core.search.indexing.ReadWriteMonitor;
+import org.eclipse.jdt.ls.core.internal.JavaLanguageServerPlugin;
+import org.eclipse.jdt.ls.core.internal.ProjectUtils;
+
+public class IndexUtils {
+    public static void copyIndexesToSharedLocation() {
+        IndexManager indexManager = JavaModelManager.getIndexManager();
+        // common index location for all workspaces
+        final String SHARED_INDEX_LOCATION = System.getProperty("jdt.core.sharedIndexLocation");
+        if (indexManager == null || StringUtils.isBlank(SHARED_INDEX_LOCATION)) {
+            return;
+        }
+
+        Set<IndexLocation> resolvedIndexLocations = new HashSet<>();
+        IPath javaCoreStateLocation = JavaCore.getPlugin().getStateLocation();
+        for (IJavaProject javaProject : ProjectUtils.getJavaProjects()) {
+            try {
+                IClasspathEntry[] entries = ((JavaProject) javaProject).getResolvedClasspath();
+                for (IClasspathEntry entry : entries) {
+                    if (entry instanceof ClasspathEntry && entry.getEntryKind() == IClasspathEntry.CPE_LIBRARY) {
+                        URL indexURL = ((ClasspathEntry) entry).getLibraryIndexLocation();
+                        if (indexURL == null) {
+                            continue;
+                        }
+
+                        IndexLocation sharedIndexLocation = IndexLocation.createIndexLocation(indexURL);
+                        IndexLocation localIndexLocation = getLocalIndexLocation(javaCoreStateLocation, entry.getPath());
+                        if (sharedIndexLocation.equals(localIndexLocation) || !resolvedIndexLocations.add(localIndexLocation)) {
+                            continue;
+                        }
+
+                        boolean needDeleteLocalIndex = false;
+                        File localIndexFile = localIndexLocation.getIndexFile();
+                        if (localIndexLocation.exists()) {
+                            File sharedIndexFile = sharedIndexLocation.getIndexFile();
+                            if (sharedIndexFile == null || localIndexFile == null) {
+                                continue;
+                            }
+
+                            if (indexManager.getIndex(sharedIndexLocation) != null) {
+                                // current classpath entry is using the shared index, delete local index directly.
+                                needDeleteLocalIndex = true;
+                            } else {
+                                Index localIndex = indexManager.getIndex(localIndexLocation);
+                                if (localIndex == null) {
```

Comment:
It might happen that after project imported, the index manager has not setup the index locations, then we skip a lot of local index.

Reply:
I think the job to write out the indices is done in the background, so you can definitely have a case where for large libraries, where they are still being written to disk after import. This is one of the reasons why in Eclipse (and likely VS Code) you can try to search after the project is imported and not get any results.

Would something like https://github.com/eclipse/eclipse.jdt.ls/blob/250ac4ff3d169ad56bf68bf6cb7bb48ef181be1f/org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/JobHelpers.java#L242 help ?

Reply:
`waitUntilIndexesReady` seems to work quite well. It can ensure all indexes are generated before copying.